### ISSUE TITLE
send intercept may cause false positive if send buffer is freed handling reply

### ISSUE THREAD

**Issue by engelsanchez** [google/sanitizers#1498]
In our internal project I noticed it's possible for the address sanitizer to report a false positive on a buffer used in a call to send() because it's freed in the reply handler for the message we are sending. I wonder if it would make sense to do the read range check on the buffer before calling the real send() to avoid any side effects of sending this buffer? This makes the address sanitizer report many false positives since our messaging system will free up request resources when the replies come in. Scheduling can make the asan check happen after that relatively easy if the message is to the same process.

in sanitizer_common_interceptors.inc:

#if SANITIZER_INTERCEPT_SEND_SENDTO
INTERCEPTOR(SSIZE_T, send, int fd, void *buf, SIZE_T len, int flags) {
  void *ctx;
  COMMON_INTERCEPTOR_ENTER(ctx, send, fd, buf, len, flags);
  if (fd >= 0) {
    COMMON_INTERCEPTOR_FD_ACCESS(ctx, fd);
    COMMON_INTERCEPTOR_FD_RELEASE(ctx, fd);
  }
  SSIZE_T res = REAL(send)(fd, buf, len, flags);
  if (common_flags()->intercept_send && res > 0)
    COMMON_INTERCEPTOR_READ_RANGE(ctx, buf, Min((SIZE_T)res, len));
  return res;
}

**Comment by vitalybuka** [comment]
We need res to know read bytes. Checking entire len for a large buffer which may need multiple sends may make quadratic algorithm from linear.
Also seems unconventional to assume how the buffer was allocated by "relpy handler".

**Comment by eugenis** [comment]
To be honest, this sounds a little scary. You are deallocating a buffer
that has been passed to a system call ("send") before that system call is
complete. This may work with the current kernel implementation, but can you
point to any documentation saying that this is part of the send() API
contract?

**Comment by engelsanchez** [comment]
@eugenis, the deallocation happens AFTER send() is called. But the address sanitizer only runs the check on the buffer AFTER send() has been called, which is my point. If the order is changed and the address sanitizer checks that the send buffer is good before calling send(), the deallocation that happens after send() will not trigger a false positive in the address sanitizer range check.

**Comment by eugenis** [comment]
On Mon, Mar 21, 2022 at 8:09 AM Engel A. Sanchez ***@***.***>
wrote:

> @eugenis <https://github.com/eugenis>, the deallocation happens AFTER
> send() is called.
>
Does it happen BEFORE send() returns though? I do not think it is safe.

But the address sanitizer only runs the check on the buffer AFTER send()

**Comment by eugenis** [comment]
It sounds like an obvious bug to me. To repeat my reasoning:

You are deallocating a buffer that has been passed to a system call
("send") before that system call is complete. This may work with the
current kernel implementation, but can you point to any documentation
saying that this is part of the send() API contract?

**Comment by rgacogne** [comment]
You are right, of course, nothing guarantees that some implementation of `send` is not going to re-read the data after sending the UDP datagram. So the ideal way would be for us to use locking to make sure this never happens, but unfortunately even an uncontended futex is not free, so we would likely get a performance regression feedback from users that are currently handling million of QPS without any issue, and that's going to be a tough sell because the problem is not going to happen in practice with any if the implementations we support.
I love ASAN and TSAN too much to remove them from our CI, so I guess I'll end up implementing locking only when we are running with ASAN or TSAN enabled, which is not great because it will likely hide potential issues, but I'm running out of ideas :-/
